File: src/content_api/handler.py
"""API Gateway HTTP API v2 Lambda entry point."""

from __future__ import annotations

import base64
import binascii
import json
import os

from .repository import DynamoContentRepository
from .service import ContentError, ContentService


def _response(status: int, body: dict) -> dict:
    return {"statusCode": status, "headers": {"content-type": "application/json"},
            "body": json.dumps(body, separators=(",", ":"))}


def _body(event: dict) -> dict:
    raw = event.get("body")
    if raw is None:
        raise ContentError("JSON body is required")
    if event.get("isBase64Encoded"):
        try:
            raw = base64.b64decode(raw, validate=True).decode("utf-8")
        except (ValueError, UnicodeError, binascii.Error) as error:
            raise ContentError("body must be valid base64-encoded UTF-8") from error
    try:
        value = json.loads(raw)
    except (ValueError, TypeError) as error:
        raise ContentError("body must be valid JSON") from error
    if not isinstance(value, dict):
        raise ContentError("body must be an object")
    return value


def lambda_handler(event: dict, _context, service: ContentService | None = None) -> dict:
    try:
        method = event.get("requestContext", {}).get("http", {}).get("method")
        route = event.get("routeKey")
        if route not in {"POST /content", "GET /content/{id}", "PUT /content/{id}"}:
            return _response(404, {"error": "route not found"})
        if service is None:
            service = ContentService(DynamoContentRepository(os.environ["TABLE_NAME"]))
        if method == "POST" and route == "POST /content":
            return _response(201, service.create(_body(event)))
        params = event.get("pathParameters") or {}
        query = event.get("queryStringParameters") or {}
        entry_id = params.get("id")
        tenant_id, site_id = query.get("tenantId"), query.get("siteId")
        if method == "GET" and route == "GET /content/{id}":
            return _response(200, service.get(entry_id, tenant_id, site_id))
        if method == "PUT" and route == "PUT /content/{id}":
            return _response(200, service.update(entry_id, tenant_id, site_id, _body(event)))
        return _response(404, {"error": "route not found"})
    except ContentError as error:
        return _response(error.status_code, {"error": str(error)})

File: src/content-api/handler.js
import { DynamoContentRepository } from "./repository.js";
import { ContentError, ContentService } from "./service.js";

let defaultService;

function response(statusCode, body) {
  return {
    statusCode,
    headers: { "content-type": "application/json" },
    body: JSON.stringify(body),
  };
}

function body(event) {
  if (event.body === undefined || event.body === null) {
    throw new ContentError("JSON body is required");
  }
  let raw = event.body;
  if (event.isBase64Encoded) {
    try {
      raw = Buffer.from(raw, "base64").toString("utf8");
    } catch {
      throw new ContentError("body must be valid base64-encoded UTF-8");
    }
  }
  try {
    const parsed = JSON.parse(raw);
    if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) {
      throw new ContentError("body must be an object");
    }
    return parsed;
  } catch (error) {
    if (error instanceof ContentError) throw error;
    throw new ContentError("body must be valid JSON");
  }
}

function service() {
  defaultService ??= new ContentService(new DynamoContentRepository(process.env.TABLE_NAME));
  return defaultService;
}

export function createHandler(injectedService = null) {
  return async function contentApiHandler(event) {
    try {
      const method = event.requestContext?.http?.method;
      const route = event.routeKey;
      if (!["POST /content", "GET /content/{id}", "PUT /content/{id}"].includes(route)) {
        return response(404, { error: "route not found" });
      }
      const contentService = injectedService ?? service();
      if (method === "POST" && route === "POST /content") {
        return response(201, await contentService.create(body(event)));
      }
      const entryId = event.pathParameters?.id;
      const tenantId = event.queryStringParameters?.tenantId;
      const siteId = event.queryStringParameters?.siteId;
      if (method === "GET" && route === "GET /content/{id}") {
        return response(200, await contentService.get(entryId, tenantId, siteId));
      }
      if (method === "PUT" && route === "PUT /content/{id}") {
        return response(200, await contentService.update(entryId, tenantId, siteId, body(event)));
      }
      return response(404, { error: "route not found" });
    } catch (error) {
      if (error instanceof ContentError) {
        return response(error.statusCode, { error: error.message });
      }
      throw error;
    }
  };
}

export const handler = createHandler();
